protocols/randhound/randhound.go

```diff
@@ -1,251 +1,995 @@
 	if err != nil {
 		return err
 	}
-	rh.GID = gid
-	rh.Group = group
 
-	// Setup session
-	session, sid, err := rh.newSession(rh.ServerIdentity().Public, purpose, time.Now())
+	if !bytes.Equal(t.SID, sid) {
+		return fmt.Errorf("Wrong session identifier")
+	}
+
+	// Verify I1 signatures
+	for _, i1 := range t.I1s {
+		if err := verifySchnorr(suite, t.CliKey, i1); err != nil {
+			return err
+		}
+	}
+
+	// Verify R1 signatures
+	for src, r1 := range t.R1s {
+		var key abstract.Point
+		for i := range t.Group {
+			for j := range t.Group[i] {
+				if src == t.Group[i][j] {
+					key = t.Key[i][j]
+				}
+			}
+		}
+		if err := verifySchnorr(suite, key, r1); err != nil {
+			return err
+		}
+	}
+
+	// Verify I2 signatures
+	for _, i2 := range t.I2s {
+		if err := verifySchnorr(suite, t.CliKey, i2); err != nil {
+			return err
+		}
+	}
+
+	// Verify R2 signatures
+	for src, r2 := range t.R2s {
+		var key abstract.Point
+		for i := range t.Group {
+			for j := range t.Group[i] {
+				if src == t.Group[i][j] {
+					key = t.Key[i][j]
+				}
+			}
+		}
+		if err := verifySchnorr(suite, key, r2); err != nil {
+			return err
+		}
+	}
+
+	// Verify message hashes HI1 and HI2; it is okay if some messages are
+	// missing as long as there are enough to reconstruct the chosen secrets
+	for i, msg := range t.I1s {
+		for _, j := range t.Group[i] {
+			if _, ok := t.R1s[j]; ok {
+				if err := verifyMessage(suite, msg, t.R1s[j].HI1); err != nil {
+					return err
+				}
+			} else {
+				log.Lvlf2("Couldn't find R1 message of server %v", j)
+			}
+		}
+	}
+
+	for i, msg := range t.I2s {
+		if _, ok := t.R2s[i]; ok {
+			if err := verifyMessage(suite, msg, t.R2s[i].HI2); err != nil {
+				return err
+			}
+		} else {
+			log.Lvlf2("Couldn't find R2 message of server %v", i)
+		}
+	}
+
+	// Verify that all servers received the same client commitment
+	for server, msg := range t.I2s {
+		for i := range msg.ChosenSecret {
+			for j := range msg.ChosenSecret[i] {
+				if int(msg.ChosenSecret[i][j]) != t.ChosenSecret[i][j] {
+					return fmt.Errorf("Server %v received wrong client commitment", server)
+				}
+			}
+		}
+	}
+
+	H, _ := suite.Point().Pick(nil, suite.Cipher(t.SID))
+	rnd := suite.Point().Null()
+	for i, group := range t.ChosenSecret {
+
+		for _, src := range group {
+
+			var poly [][]byte
+			var encPos []int
+			var encShare []abstract.Point
+			var encProof []ProofCore
+			var X []abstract.Point
+
+			var decPos []int
+			var decShare []abstract.Point
+			var decProof []ProofCore
+
+			// All R1 messages of the chosen secrets should be there
+			if _, ok := t.R1s[src]; !ok {
+				return errors.New("R1 message not found")
+			}
+			r1 := t.R1s[src]
+
+		OUTER:
+			for j := 0; j < len(r1.EncShare); j++ {
+
+				// Check availability of corresponding R2 messages, skip if not there
+				target := r1.EncShare[j].Target
+				if _, ok := t.R2s[target]; !ok {
+					continue OUTER
```

Comment:
Note that the r1 message changes the next time you go to `OUTER`, see https://github.com/dedis/cothority/blob/randhound_341/protocols/randhound/randhound.go#L366.